src/Bicep.Core/TypeSystem/OperationReturnTypeEvaluator.cs

```diff
@@ -1,180 +1,267 @@
-        new(BinaryOperator.GreaterThanOrEqual, LanguageConstants.Int, LanguageConstants.Int, "greaterOrEquals", LanguageConstants.Bool),
+        new LessThanEvaluator(),
+        new LessThanOrEqualToEvaluator(),
+        new GreaterThanEvaluator(),
+        new GreaterThanOrEqualToEvaluator(),
 
         // relational (string)
-        new(BinaryOperator.LessThan, LanguageConstants.String, LanguageConstants.String, "less", LanguageConstants.Bool),
-        new(BinaryOperator.LessThanOrEqual, LanguageConstants.String, LanguageConstants.String, "lessOrEquals", LanguageConstants.Bool),
-        new(BinaryOperator.GreaterThan, LanguageConstants.String, LanguageConstants.String, "greater", LanguageConstants.Bool),
-        new(BinaryOperator.GreaterThanOrEqual, LanguageConstants.String, LanguageConstants.String, "greaterOrEquals", LanguageConstants.Bool),
+        new BinaryEvaluator(BinaryOperator.LessThan, (LanguageConstants.String, LanguageConstants.String), "less", LanguageConstants.Bool),
+        new BinaryEvaluator(BinaryOperator.LessThanOrEqual, (LanguageConstants.String, LanguageConstants.String), "lessOrEquals", LanguageConstants.Bool),
+        new BinaryEvaluator(BinaryOperator.GreaterThan, (LanguageConstants.String, LanguageConstants.String), "greater", LanguageConstants.Bool),
+        new BinaryEvaluator(BinaryOperator.GreaterThanOrEqual, (LanguageConstants.String, LanguageConstants.String), "greaterOrEquals", LanguageConstants.Bool),
 
         // arithmetic
-        new(BinaryOperator.Add, LanguageConstants.Int, LanguageConstants.Int, "add", LanguageConstants.Int),
-        new(BinaryOperator.Subtract, LanguageConstants.Int, LanguageConstants.Int, "sub", LanguageConstants.Int),
-        new(BinaryOperator.Multiply, LanguageConstants.Int, LanguageConstants.Int, "mul", LanguageConstants.Int),
-        new(BinaryOperator.Divide, LanguageConstants.Int, LanguageConstants.Int, "mul", LanguageConstants.Int),
-        new(BinaryOperator.Modulo, LanguageConstants.Int, LanguageConstants.Int, "mod", LanguageConstants.Int),
+        new AdditionEvaluator(),
+        new SubtractionEvaluator(),
+        new BinaryEvaluator(BinaryOperator.Multiply, (LanguageConstants.Int, LanguageConstants.Int), "mul", LanguageConstants.Int),
+        new BinaryEvaluator(BinaryOperator.Divide, (LanguageConstants.Int, LanguageConstants.Int), "div", LanguageConstants.Int),
+        new ModuloEvaluator(),
 
         // coalesce
-        new(BinaryOperator.Coalesce, LanguageConstants.Any, LanguageConstants.Any, "coalesce", LanguageConstants.Any),
-    }.ToImmutableArray();
+        new CoalesceEvaluator());
 
-    internal static TypeSymbol? FoldUnaryExpression(UnaryOperationSyntax expressionSyntax, TypeSymbol operandType, out IEnumerable<IDiagnostic> foldDiagnostics)
+    public static TypeSymbol? TryFoldUnaryExpression(UnaryOperationSyntax expressionSyntax, TypeSymbol operandType)
```

Comment:
How come `foldDiagnostics` was removed from `TryFoldUnaryExpression()`?

Reply:
The `foldDiagnostics` output parameter is basically only there because invoking the ARM expressions engine might throw an exception, and we want to convert that to a warning. Unary expression handling no longer invokes the ARM engine -- it's all based on type analysis -- so there are no diagnostics that could be raised.

I could add `foldDiagnostics` back if we want to retain the flexibility to emit warning diagnostics here, but I think maybe passing in an `IDiagnosticsWriter` is probably a better design. Since this was an `internal` method, the API is not as elegant as I would have liked. Maybe that should be addressed now that these methods have been made `public`?

Reply:
Updated `TryFoldUnaryExpression` and `TryFoldBinaryExpression` to accept an `IDiagnosticWriter` in a711cad.